tests/CredentialsLoaderTest.php

```diff
@@ -29,6 +29,123 @@ public function testUpdateMetadataSkipsWhenAuthenticationisSet()
         $this->assertArrayHasKey('authentication', $metadata);
         $this->assertEquals('foo', $metadata['authentication']);
     }
+
+    /** @runInSeparateProcess */
+    public function testGetDefaultClientCertSource()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/valid');
+
+        $callback = CredentialsLoader::getDefaultClientCertSource();
+        $this->assertNotNull($callback);
+
+        $output = $callback();
+        $this->assertEquals('foo', $output);
+    }
+
+    /** @runInSeparateProcess */
+    public function testNonExistantDefaultClientCertSource()
+    {
+        putenv('HOME=');
+
+        $callback = CredentialsLoader::getDefaultClientCertSource();
+        $this->assertNull($callback);
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage Invalid client cert source JSON
+     */
+    public function testDefaultClientCertSourceInvalidJsonThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidjson');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage cert source requires "cert_provider_command"
+     */
+    public function testDefaultClientCertSourceInvalidKeyThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidkey');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     * @expectedException UnexpectedValueException
+     * @expectedExceptionMessage cert source expects "cert_provider_command" to be an array
+     */
+    public function testDefaultClientCertSourceInvalidValueThrowsException()
+    {
+        putenv('HOME=' . __DIR__ . '/fixtures4/invalidvalue');
+
+        CredentialsLoader::getDefaultClientCertSource();
+    }
+
+    /**
+     * @runInSeparateProcess
+     */
+    public function testActualDefaultClientCertSource()
+    {
+        $clientCertSource = CredentialsLoader::getDefaultClientCertSource();
+        if (is_null($clientCertSource)) {
+            $this->markTestSkipped('No client cert source found');
```

Comment:
What are the cases where this wouldn't be found? Should we stub it out?

Reply:
The  client cert source is only found when the corresponding file exists (`~/.secureConnect/context_aware_metadata.json`). This is tested in `testNonExistantDefaultClientCertSource`!